internal/controlplane/handlers_oauth.go

```diff
@@ -115,55 +127,93 @@ func (s *Server) GetAuthorizationURL(ctx context.Context,
+		ctx := context.Background()
+
+		if err := s.processCallback(ctx, w, r, pathParams); err != nil {
+			log.Printf("error handling provider callback: %s", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+}
+
+func (s *Server) processCallback(ctx context.Context, w http.ResponseWriter, r *http.Request,
+	pathParams map[string]string) error {
+	state := r.URL.Query().Get("state")
+	code := r.URL.Query().Get("code")
 
 	// Configure tracing
 	span := trace.SpanFromContext(ctx)
-	span.SetName("server.ExchangeCodeForTokenCLI")
-	span.SetAttributes(attribute.Key("code").String(in.Code))
+	span.SetName("server.HandleProviderCallback")
+	span.SetAttributes(attribute.Key("code").String(code))
 	defer span.End()
 
-	// Check the nonce to make sure it's valid
-	valid, err := mcrypto.IsNonceValid(in.State, s.cfg.Auth.NoncePeriod)
+	provider := pathParams["provider"]
 
+	// Check the nonce to make sure it's valid
+	valid, err := mcrypto.IsNonceValid(state, s.cfg.Auth.NoncePeriod)
 	if err != nil {
-		return nil, status.Errorf(codes.Unknown, "error checking nonce: %s", err)
+		return fmt.Errorf("error checking nonce: %w", err)
 	}
-
 	if !valid {
-		return nil, status.Error(codes.InvalidArgument, "invalid nonce")
+		return errors.New("invalid nonce")
 	}
 
 	// get projectID from session along with state nonce from the database
-	stateData, err := s.store.GetProjectIDPortBySessionState(ctx, in.State)
+	stateData, err := s.store.GetProjectIDPortBySessionState(ctx, state)
 	if err != nil {
-		return nil, status.Errorf(codes.Unknown, "error getting project ID by session state: %s", err)
+		return fmt.Errorf("error getting project ID by session state: %w", err)
 	}
 
-	// get provider
-	provider, err := getProviderFromRequestOrDefault(ctx, s.store, in, stateData.ProjectID)
+	err = s.generateOAuthToken(ctx, provider, code, stateData)
 	if err != nil {
-		return nil, providerError(err)
+		return err
 	}
 
+	// Telemetry logging
+	logger.BusinessRecord(ctx).Provider = provider
+	logger.BusinessRecord(ctx).Project = stateData.ProjectID
+
+	if stateData.RedirectUrl.Valid {
+		parsedURL, err := url.Parse(stateData.RedirectUrl.String)
+		if err != nil {
+			return fmt.Errorf("error parsing redirect URL: %w", err)
+		}
+		http.Redirect(w, r, parsedURL.String(), http.StatusTemporaryRedirect)
+		if err != nil {
+			return fmt.Errorf("error redirecting to URL: %w", err)
+		}
+	} else {
+		w.Header().Set("Content-Type", "text/html; charset=utf-8")
+		_, err = w.Write(auth.OAuthSuccessHtml)
+		if err != nil {
+			return fmt.Errorf("error writing OAuth success page: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (s *Server) generateOAuthToken(ctx context.Context, provider string, code string,
+	stateData db.GetProjectIDPortBySessionStateRow) error {
 	// generate a new OAuth2 config for the given provider
-	oauthConfig, err := auth.NewOAuthConfig(provider.Name, true)
+	oauthConfig, err := auth.NewOAuthConfig(provider, true)
```

Comment:
We used to use `getProviderFromRequestOrDefault`, but now we're just taking the path parameter as-is and feeding it in to `NewOAuthConfig`.  I think we need to do some validation here.

Reply:
We should check that the provider is not empty, but `getProviderFromRequestOrDefault` never made sense here because this endpoint was only ever called from GitHub which never had a user context.

Reply:
I agree